core/embed/trezorhal/stm32f4/syscall_verifiers.c

```diff
@@ -0,0 +1,667 @@
+  message_len = message ? MIN(message_len, sizeof(message_copy) - 1) : 0;
+  footer_len = footer ? MIN(footer_len, sizeof(footer_copy) - 1) : 0;
+
+  if (title != NULL) {
+    if (!probe_read_access(title, title_len)) {
+      goto access_violation;
+    }
+
+    title = strcpy(title_copy, title);
+  }
+
+  if (message != NULL) {
+    if (!probe_read_access(message, message_len)) {
+      goto access_violation;
+    }
+
+    message = strcpy(message_copy, message);
+  }
+
+  if (footer != NULL) {
+    if (!probe_read_access(footer, footer_len)) {
+      goto access_violation;
+    }
+
+    footer = strcpy(footer_copy, footer);
+  }
+
+  systask_exit_error(task, title, title_len, message, message_len, footer,
+                     footer_len);
+
+  return;
+
+access_violation:
+  apptask_access_violation();
+}
+
+void systask_exit_fatal__verified(systask_t *task, const char *message,
+                                  size_t message_len, const char *file,
+                                  size_t file_len, int line) {
+  char message_copy[64] = {0};
+  char file_copy[64] = {0};
+
+  message_len = message ? MIN(message_len, sizeof(message_copy) - 1) : 0;
+  file_len = file ? MIN(file_len, sizeof(file_copy) - 1) : 0;
+
+  if (message != NULL) {
+    if (!probe_read_access(message, message_len)) {
+      goto access_violation;
+    }
+
+    message = strcpy(message_copy, message);
+  }
+
+  if (file != NULL) {
+    if (!probe_read_access(file, file_len)) {
+      goto access_violation;
+    }
+
+    file = strcpy(file_copy, file);
+  }
+
+  systask_exit_fatal(task, message, message_len, file, file_len, line);
+
+  return;
+
+access_violation:
+  apptask_access_violation();
+}
+
+// ---------------------------------------------------------------------
+
+void reboot_and_upgrade__verified(const uint8_t hash[32]) {
+  if (!probe_read_access(hash, 32)) {
+    goto access_violation;
+  }
+
+  reboot_and_upgrade(hash);
+
+access_violation:
+  apptask_access_violation();
+}
+
+// ---------------------------------------------------------------------
+
+#ifdef XFRAMEBUFFER
+
+bool display_get_frame_buffer__verified(display_fb_info_t *fb) {
+  if (!probe_write_access(fb, sizeof(*fb))) {
+    goto access_violation;
+  }
+
+  display_fb_info_t fb_copy = {0};
+
+  bool result = display_get_frame_buffer(&fb_copy);
+
+  *fb = fb_copy;
+
+  return result;
+
+access_violation:
+  apptask_access_violation();
+  return false;
+}
+
+#endif  // XFRAMEBUFFER
+
+void display_fill__verified(const gfx_bitblt_t *bb) {
+  if (!probe_read_access(bb, sizeof(*bb))) {
+    goto access_violation;
+  }
+
+  display_fill(bb);
+
+  return;
+
+access_violation:
+  apptask_access_violation();
+}
+
+void display_copy_rgb565__verified(const gfx_bitblt_t *bb) {
+  if (!probe_read_access(bb, sizeof(*bb))) {
+    goto access_violation;
+  }
+
+  uint8_t *src_ptr = (uint8_t *)bb->src_row;
+  size_t src_len = bb->src_stride * bb->height;
+
+  if (!probe_read_access(src_ptr, src_len)) {
```

Comment:
I think the checks here are not sufficient: 

- the bb->width parameter is not checked which can lead to reading data beyond allowed area, regardless of src_stride.
- i don't think we can ignore the write access here. I.e. the bitblt sets the dst_y and thus can easily cause an overflow of designated area, and looking into for example st7789 implementation with framebuffer, this is not checked in the driver - nor do i think that this should be delegated only to the driver. I realize this is tricky as sometimes we write into register and sometimes to framebuffer, but we should probably do some clipping here to ensure that these functions can only ask to write into the expected display area. 

(applies to fill function too)

Reply:
Yes, I’m aware of this – I wanted to postpone it a bit (not incorporate it into this PR) since it needs to be addressed in the implementation of the function within the driver itself. Only the driver has enough information to validate the destination parameters (for example, it currently checks the display resolution, it knows if we are using framebuffer or not...).

I wouldn’t handle clipping here (as it’s quite complex and is already managed in the Rust layer). I’d prefer to perform simple checks here to ensure we are not writing/reading beyond the unprivileged area.

Let’s solve it a bit later, and I’ll create a new issue.

Reply:
Added the issue https://github.com/trezor/trezor-firmware/issues/4273